Add FastAPI framework as next step for API framework

mvdbeek opened the pull request:
[FastAPI](https://fastapi.tiangolo.com/) does a lot of things (websockets, graphql, Oauth2 + JWT tokens, backend tasks, celery integration, sync + async routes, openapi + typing, good performance, file serving, active community and nice documentation) and can be integrated nicely route-by-route without having to toss out everything at once.

This is a quick experiment where I replaced /api/jobs, IFF you start galaxy with `python scripts/fapi.py -c config/galaxy.yml` (need to activate the virtualenv and install uvicorn). The rest of routes continues to work as normal, so Galaxy is fully functional.

Here's the openAPI doc for the new route:

<img width="1785" alt="Screenshot 2020-11-30 at 10 20 24" src="https://user-images.githubusercontent.com/6804901/100591316-dc570980-32f5-11eb-9743-ed9ec9ac0b36.png">

In terms of challenges:
  - I don't think uwsgi can serve ASGI apps
  - We'd have to replace the interactive tool routing
  - We'd have to get buy-in from admins to abandon uwsgi


TODO for this PR:
- [x] [test API with paste *and* uvicorn until we drop (u)wsgi-only support](3ab8a027ecac9b67ac5ba753fb1f0a1fa2fcb8be)
- [x] [fix job lock location](7d3c6db3ce7901e80ced461575d151e199b74cd2)

mvdbeek committed:
Expose api/jobs/show via FastAPI

mvdbeek committed:
Adjust galaxy-main script so we can import from it for easy app setup

mvdbeek committed:
Use InferringRouter to infer response model from type annotation

mvdbeek committed:
Add fastapi requirements

mvdbeek committed:
Fix lru_cache decorator

nsoranzo commented on scripts/galaxy_main.py:
```diff
@@ -0,0 +1 @@
+galaxy-main
```
Maybe `git mv scripts/galaxy-main scripts/galaxy_main.py` and symlink `galaxy-main` to `galaxy_main.py` instead?

mvdbeek replied:
Done that in a rebase!

nsoranzo commented on scripts/fapi.py:
```diff
@@ -0,0 +1,32 @@
+import uvicorn
```
Need to add `uvicorn` to the requirements?

mvdbeek replied:
It's not needed for startup, so as a development-only thing we'd probably move it in dev-requirements ?

mvdbeek committed:
Structure FastAPI dependencies

For now dump them into a single module, but that could become a package
of course.

mvdbeek committed:
Use added managers in old GalaxyWebTransaction

nsoranzo commented on lib/galaxy/webapps/base/webapp.py:
```diff
@@ -23,7 +23,9 @@
 
 from galaxy import util
-from galaxy.exceptions import ConfigurationError, MessageException
+from galaxy.exceptions import AuthenticationFailed, ConfigurationError, MessageException
```
```suggestion
from galaxy.exceptions import (
    AuthenticationFailed,
    ConfigurationError,
    MessageException,
)
```

mvdbeek replied:
Added that in a rebase, thanks!

mvdbeek committed:
Add JobLock schema, validate encoded database id with custom field

nsoranzo commented on lib/galaxy/webapps/galaxy/api/dependencies.py:
```diff
@@ -0,0 +1,91 @@
+"""
+This module *does not* contain API routes. It exclusively contains dependencies to be used in FastAPI routes
```
Maybe this stuff can go in `lib/galaxy/webapps/galaxy/api/__init__.py` ? That would be less confusing, I think.

mvdbeek replied:
Yep, done that for now. I suppose this will grow significantly, but I think that's a good location for now.

mvdbeek committed:
Move dependencies into galaxy.webapps.galaxy.api

mvdbeek committed:
Pass model into GalaxySessionManager

This is a quick and dirty way to pass in the correct model. Would seem
cleaner to parameterize this based on whether this is a Galaxy app or a
Toolshed app.

mvdbeek committed:
Don't fail if there's no galaxysession

mvdbeek committed:
Add uvicorn to dependencies

mvdbeek committed:
Update all dependencies

mvdbeek committed:
Work around pygithub pin

jmchilton committed:
Tests cases with uvicorn.

jmchilton committed:
Outline how to share complex code between old and new style controllers.

More use of Pydantic models in the old code and more use of "view" functions below the controller code - mirroring inn some ways what is done for dependency management APIs (which I had structured to be very lightweight in an alternative framework such as this).

mvdbeek committed:
Fix view_show_job call

mvdbeek committed:
Acquire event loop in thread

mvdbeek committed:
Make WSGI headers strings

jmchilton committed:
Migrate API role to modern Galaxy architecture - pydantic, galaxy.exceptions, managers.

Transition /api/roles/{id} to return a single role and not a singleton list (breaking change but better API).

mvdbeek committed:
Don't compare by object identity

mvdbeek committed:
Don't use lru_cache for now

mvdbeek committed:
Replace api_inherit_admin with BootstrapAdminUser

mvdbeek committed:
Use list comprehension

mvdbeek committed:
Adjust test for singular role creation response

mvdbeek committed:
Is that the problem for the expired API key issue

mvdbeek committed:
get rid of more lru_cache

mvdbeek committed:
Get a new trans on each request

mvdbeek committed:
Fix role unit test

Previous MockApp user was implcitly admin (no user, no session).

mvdbeek committed:
Add pydantic to web_framework requirements

mvdbeek committed:
Expunge all

mvdbeek committed:
Refresh user

davelopez commented on lib/galaxy/webapps/galaxy/api/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    return UserManager(app)
+
+
+def get_session_manager(app=Depends(get_app)) -> GalaxySessionManager:
```
Nice to see some type annotations kicking in! :)
```suggestion
def get_session_manager(app: UniverseApplication = Depends(get_app)) -> GalaxySessionManager:
```

davelopez commented on lib/galaxy/webapps/galaxy/api/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    return app.model.session
+
+
+def get_user_manager(app: UniverseApplication = Depends(get_app)):
```
```suggestion
def get_user_manager(app: UniverseApplication = Depends(get_app)) -> UserManager:
```

davelopez commented on lib/galaxy/webapps/galaxy/api/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    return api_user
+
+
+def get_trans(app=Depends(get_app), user: Optional[User] = Depends(get_user),
```
```suggestion
def get_trans(app: UniverseApplication = Depends(get_app), user: Optional[User] = Depends(get_user),
```

davelopez commented on lib/galaxy/webapps/galaxy/api/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    return SessionRequestContext(app=app, user=user, galaxy_session=galaxy_session)
+
+
+def get_admin_user(trans: SessionRequestContext = Depends(get_trans), user_manager: UserManager = Depends(get_user_manager)):
```
I'm not 100% sure if the return type can be `User` or if there is some specific `AdminUser` class or something like that.
```suggestion
def get_admin_user(trans: SessionRequestContext = Depends(get_trans), user_manager: UserManager = Depends(get_user_manager)) -> Optional[User]:
```

mvdbeek replied:
There is a `BooststrapAdminUser` now, so this could be a typing.Union[User, BootstrapAdminUser].
I'm not 100% convinced it makes sense to annotate everything, since in this case I guess we'd have to import the `BooststrapAdminUser` just to create the type annotation. Which I think is fine in this case, but I could see this becoming a burden in other cases.

davelopez replied:
This is really a good point. I haven't thought of this burden because I have used them only in a small project.
Should we come up with some kind of Type Annotation Guidelines for Galaxy? or are there any rules or best practices available for this matter?

mvdbeek replied:
Yes, that is a good idea, but maybe we can freestyle it a bit while we're still getting to know the pros and cons?
My rough thinking is that for everything that we consider public (like API routes) we should annotate all parameters and return types. For things that are really dynamic we might make it optional ? For functions that are meant for re-use, try to include type annotations ?

davelopez replied:
That makes perfect sense. Totally agree!

jmchilton replied:
I think we should probably annotate an alternative admin transaction so we don't need to inject an admin user that isn't used? Or just reuse the old decorator to be honest.

One thing is we're annotating explicit concrete classes for everything and it is causing me some distress - we're creating very explicit dependencies and deep transitive dependencies. If we're going to use types internally for things like app and trans we should be using more precise interfaces that they implement and that can be imported without bringing all of say app's dependencies.

davelopez reviewed (commented):
Love to see type hints spreading, they make my IDE and me happy :)

jmchilton committed:
Handle pydantic validation errors as galaxy exceptions, migrate roles.

jmchilton committed:
Move role functionality into RolesManager.

jmchilton committed:
Add FastAPI roles controller, de-duplicate app initialization for FastAPI.

jmchilton committed:
Unify exception handling across FastAPI and wsgi.

Also use JSON requests for roles tests instead of form-data - not sure how we're going to tackle that.

jmchilton committed:
Lint and spelling fixes.

jmchilton committed:
Fix for roles API changeover.

mvdbeek committed:
Add more type annotations

Thanks @davelopez!

Co-authored-by: davelopez <46503462+davelopez@users.noreply.github.com>

jmchilton commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -30,10 +43,62 @@
+    hda_manager: hdas.HDAManager = Depends(get_hda_manager)
+
+    @router.get('/job_lock')
+    def job_lock_status(self, admin_user=Depends(get_admin_user)) -> JobLock:
```
Is it intentional that this has a different route than the existing controller - because I think the existing route is api/job_lock not api/jobs/job_lock?

mvdbeek replied:
No, just an oversight.

jmchilton committed:
Sync admin checking across old and new API frameworks.

jmchilton committed:
I guess we don't need that.

mvdbeek committed:
Fix trans.trans typo

mvdbeek committed:
Fix routes

Bare routes (without trailing /) don't quite work, so we need to fully
specify them. Also fixes job lock route to be at `/api/job_lock` instead
of /api/jobs/job_lock.

mvdbeek committed:
Also (temporarily) test against paste
